packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -486,8 +486,85 @@ class BuildResult {
 }
 
 /// The build system is responsible for invoking and ordering [Target]s.
-class BuildSystem {
-  const BuildSystem({
+abstract class BuildSystem {
+  /// Const constructor to allow subclasses to be const.
+  const BuildSystem();
+
+  /// Build [target] and all of its dependencies.
+  Future<BuildResult> build(
+    Target target,
+    Environment environment, {
+    BuildSystemConfig buildSystemConfig = const BuildSystemConfig(),
+  });
+
+  /// Perform an incremental build of [target] and all of its dependencies.
+  ///
+  /// If [previousBuild] is not provided, a new incremental build is
+  /// initialized.
+  Future<BuildResult> buildIncremental(
+    Target target,
+    Environment environment,
+    BuildResult previousBuild,
+  );
+
+  /// Write the identifier of the last build into the output directory and
+  /// remove the previous build's output.
+  ///
+  /// The build identifier is the basename of the build directory where
+  /// outputs and intermediaries are written, under `.dart_tool/flutter_build`.
+  /// This is computed from a hash of the build's configuration.
+  ///
+  /// This identifier is used to perform a targeted cleanup of the last output
+  /// files, if these were not already covered by the built-in cleanup. This
+  /// cleanup is only necessary when multiple different build configurations
+  /// output to the same directory.
+  @visibleForTesting
+  static void trackSharedBuildDirectory(
+    Environment environment,
+    FileSystem fileSystem,
+    Map<String, File> currentOutputs,
+  ) {
+    final String currentBuildId = fileSystem.path.basename(environment.buildDir.path);
+    final File lastBuildIdFile = environment.outputDir.childFile('.last_build_id');
+    if (!lastBuildIdFile.existsSync()) {
+      lastBuildIdFile.writeAsStringSync(currentBuildId);
```

Comment:
Hopefully this is only used for testing and build fns in this class?

If so, can it move to `FlutterBuildSystem`?

Reply:
Done